src/src/Helpers/TarStream.h

```diff
@@ -0,0 +1,175 @@
+# define TSVTX    01000 /* reserved */
+                        /* file permissions */
+# define TUREAD   00400 /* read by owner */
+# define TUWRITE  00200 /* write by owner */
+# define TUEXEC   00100 /* execute/search by owner */
+# define TGREAD   00040 /* read by group */
+# define TGWRITE  00020 /* write by group */
+# define TGEXEC   00010 /* execute/search by group */
+# define TOREAD   00004 /* read by other */
+# define TOWRITE  00002 /* write by other */
+# define TOEXEC   00001 /* execute/search by other */
+
+constexpr size_t TAR_HEADER_SIZE = sizeof(posix_header);
+constexpr size_t TAR_BLOCK_SIZE  = 512u;
+
+struct TarFileInfo_struct {
+  TarFileInfo_struct() {}
+
+  TarFileInfo_struct(const String _fileName,
+                     size_t       _fileSize);
+
+  String fileName;
+  size_t fileSize; // Actual file size in bytes
+  size_t tarSize;  // File size rounded up to the next 512 byte block (tar block-size)
+};
+
+enum TarStreamState_e : uint8_t {
+  Initial = 0u,  // Initial state
+  ReadingHeader, // Processing a file header
+  ReadingFile,   // Reading a file
+  ReadingSlack,  // Reading the slack space after the file data is processed
+  ReadingFinal,  // Reading the empty block after the last file
+  WritingHeader, // Writing the header data
+  WritingFile,   // Writing the file data
+  WritingSlack,  // Writing nothing, receiving the slack space
+  WritingFinal,  // Writing nothing, receiving the final block
+  WritingDone,   // We're done, just skip any incoming bytes
+  Error,
+};
+
+class TarStream : public Stream {
+public:
+
+  TarStream();
+  TarStream(const String fileName);
+  TarStream(const String      fileName,
+            FileDestination_e destination);
+  virtual ~TarStream();
+
+  virtual size_t      write(uint8_t ch);
+
+  virtual size_t      write(const uint8_t *buf,
+                            size_t         size);
+  virtual int         available();
+  virtual int         read();
+  virtual int         peek();
+  virtual void        flush();
+  virtual size_t      size();
+  virtual const char* name();
+
+  bool                addFileIfExists(const String& fileName); // Check file and add to list if exists
+  bool                addFile(const String& fileName,
+                              size_t        fileSize);         // Add a file to the list and update _tarSize
+  bool                isFileIncluded(const String& filename);  // Is this file included?
+  size_t              getFileCount() const;                    // Actual number of files in the achive when uploading
+  size_t              getFilesSizes() {                        // Actual size of all files in bytes
+    return _filesSizes;
+  }
+
+private:
+
+  void clearHeader();
+  void setupHeader();
+  bool validateHeader();
+
+  union {
+    posix_header _tarHeader;                                  // Header
+    uint8_t      _tarData[TAR_HEADER_SIZE];                   // Access to the tarheader data
```

Comment:
You can initialize this `_tarData` using `{}` as it will also clear the content of `_tarHeader` since it is in the same union.

Reply:
It has to be cleared multiple times during processing, so there's a private method for that, but it won't hurt to clear it initially too.